container/lxd/initialisation.go

```diff
@@ -49,6 +76,147 @@ func getPackagingConfigurer(series string) (config.PackagingConfigurer, error) {
 	return config.NewPackagingConfigurer(series)
 }
 
+func configureZFS() {
+	/* create a 100 GB pool by default (sparse, so it won't actually fill
+	 * that immediately)
+	*/
+	output, err := exec.Command(
+		"lxd",
+		"init",
+		"--auto",
+		"--storage-backend", "zfs",
+		"--storage-pool", "lxd",
+		"--storage-create-loop", "100",
+	).CombinedOutput()
+
+	if err != nil {
+		logger.Warningf("configuring zfs failed with %s: %s", err, string(output))
+	}
+}
+
+func configureLXDBridge() error {
+	f, err := os.OpenFile(lxdBridgeFile, os.O_RDWR, 0777)
+	if err != nil {
+		/* We're using an old version of LXD which doesn't have
+		 * lxd-bridge; let's not fail here.
+		 */
+		if os.IsNotExist(err) {
+			logger.Warningf("couldn't find %s, not configuring it", lxdBridgeFile)
+			return nil
+		}
+		return errors.Trace(err)
+	}
+	defer f.Close()
+
+	output, err := exec.Command("ip", "addr", "show").CombinedOutput()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	subnet, err := detectSubnet(string(output))
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	existing, err := ioutil.ReadAll(f)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	result := editLXDBridgeFile(string(existing), subnet)
+	_, err = f.Seek(0, 0)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	_, err = f.WriteString(result)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+
+	/* non-systemd systems don't have the lxd-bridge service, so this always fails */
+	_ = exec.Command("service", "lxd-bridge", "restart").Run()
+	return exec.Command("service", "lxd", "restart").Run()
+}
+
+func detectSubnet(ipAddrOutput string) (string, error) {
+	max := 0
+
+	for _, line := range strings.Split(ipAddrOutput, "\n") {
+		trimmed := strings.TrimSpace(line)
+		if trimmed == "" {
+			continue
+		}
+
+		columns := strings.Split(trimmed, " ")
+
+		if len(columns) < 2 {
+			return "", errors.Trace(fmt.Errorf("invalid ip addr output line %s", line))
+		}
+
+		if columns[0] != "inet" {
+			continue
+		}
+
+		addr := columns[1]
+		if !strings.HasPrefix(addr, "10.0.") {
+			continue
+		}
+
+		tuples := strings.Split(addr, ".")
+		if len(tuples) < 4 {
+			return "", errors.Trace(fmt.Errorf("invalid ip addr %s", addr))
+		}
+
+		subnet, err := strconv.Atoi(tuples[2])
+		if err != nil {
+			return "", errors.Trace(err)
+		}
+
+		if subnet > max {
+			max = subnet
+		}
+	}
+
+	return fmt.Sprintf("%d", max+1), nil
```

Comment:
On Mon, Mar 28, 2016 at 11:24:27PM -0700, John Arbash Meinel wrote:
> > +		tuples := strings.Split(addr, ".")
> > +		if len(tuples) < 4 {
> > +			return "", errors.Trace(fmt.Errorf("invalid ip addr %s", addr))
> > +		}
> > +
> > +		subnet, err := strconv.Atoi(tuples[2])
> > +		if err != nil {
> > +			return "", errors.Trace(err)
> > +		}
> > +
> > +		if subnet > max {
> > +			max = subnet
> > +		}
> > +	}
> > +
> > +	return fmt.Sprintf("%d", max+1), nil
> 
> I wonder if "max+1" is the best answer. I suppose it is *an* answer. But it feels like if I ever configure subnet 10.0.255.* for something, even though the other 250 or so are empty, then this will actually create a bad configuration and fail.
> I don't know what people have seen in the real world (do people generally do low numbers and fill them mostly full), or do they pick "whole" numbers like 10, 100, 200, etc. (I know when I've created local bridges for testing, I use the whole-number approach because it gives me room to play with it.)
> 
> I guess an auto-guess function should pick something that may be sane, but can't really understand your network at all. But we should really have a way to pass this in via config, so that users doing "juju bootstrap" can tell *us* something about the logic of their networking space.

We can certainly choose something else, this is just a copy of the
behavior of lxcbr0, which is what is currently happening. I figured
I'd leave something more advanced up to you guys.

> 
> ---
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/juju/juju/pull/4903/files/af7d4537ba59659bb1087fc418f257280034e6f2#r57676741